File: py/todoosy/query.py
"""
Todoosy Query Engine
"""

from dataclasses import dataclass, field
from typing import Optional

from .parser import parse
from .types import AST, ItemNode, UpcomingItem, MiscItem, Scheme


@dataclass
class UpcomingResult:
    items: list[UpcomingItem] = field(default_factory=list)


@dataclass
class MiscResult:
    items: list[MiscItem] = field(default_factory=list)


def build_path(item_id: str, ast: AST) -> str:
    """Build the path to an item."""
    item_map = {item.id: item for item in ast.items}
    parent_map: dict[str, str] = {}

    for item in ast.items:
        for child_id in item.children:
            parent_map[child_id] = item.id

    parts: list[str] = []
    current_id: Optional[str] = item_id

    while current_id is not None:
        item = item_map.get(current_id)
        if item:
            parts.insert(0, item.title_text)
        current_id = parent_map.get(current_id)

    return ' > '.join(parts)


def query_upcoming(text: str, scheme: Optional[Scheme] = None) -> UpcomingResult:
    """Get all items with due dates, sorted by date and priority."""
    result = parse(text)
    ast = result.ast
    items: list[UpcomingItem] = []

    for item in ast.items:
        if item.metadata.due:
            upcoming_item = UpcomingItem(
                id=item.id,
                due=item.metadata.due,
                priority=item.metadata.priority,
                path=build_path(item.id, ast),
                item_span=item.item_span,
            )

            # Add priority label if scheme is provided
            if scheme and item.metadata.priority is not None:
                label = scheme.priorities.get(str(item.metadata.priority))
                if label:
                    upcoming_item.priority_label = label

            items.append(upcoming_item)

    # Sort by:
    # 1. Due date ascending
    # 2. Priority ascending (lower is higher priority, None treated as infinity)
    # 3. Document order
    def sort_key(item: UpcomingItem) -> tuple:
        priority = item.priority if item.priority is not None else float('inf')
        return (item.due, priority, item.item_span[0])

    items.sort(key=sort_key)

    return UpcomingResult(items=items)


def query_misc(text: str) -> MiscResult:
    """Get all items under the Misc section."""
    result = parse(text)
    ast = result.ast
    items: list[MiscItem] = []

    # Find the Misc section
    misc_section_id = None
    for item in ast.items:
        if item.type == 'heading' and item.title_text == 'Misc' and item.level == 1:
            misc_section_id = item.id
            break

    if misc_section_id is None:
        return MiscResult(items=items)

    # Get all direct children of the Misc section
    item_map = {item.id: item for item in ast.items}
    misc_section = item_map.get(misc_section_id)

    if not misc_section:
        return MiscResult(items=items)

    for child_id in misc_section.children:
        child = item_map.get(child_id)
        if child:
            items.append(MiscItem(
                id=child.id,
                title_text=child.title_text,
                item_span=child.item_span,
            ))

    return MiscResult(items=items)

File: js/src/query.ts
/**
 * Todoosy Query Engine
 */

import { parse, type ParseOptions } from './parser.js';
import type {
  AST,
  ItemNode,
  UpcomingItem,
  UpcomingResult,
  MiscItem,
  MiscResult,
  Scheme,
} from './types.js';

function buildPath(itemId: string, ast: AST): string {
  const itemMap = new Map<string, ItemNode>();
  const parentMap = new Map<string, string>();

  for (const item of ast.items) {
    itemMap.set(item.id, item);
    for (const childId of item.children) {
      parentMap.set(childId, item.id);
    }
  }

  const parts: string[] = [];
  let currentId: string | undefined = itemId;

  while (currentId !== undefined) {
    const item = itemMap.get(currentId);
    if (item) {
      parts.unshift(item.title_text);
    }
    currentId = parentMap.get(currentId);
  }

  return parts.join(' > ');
}

export function queryUpcoming(text: string, scheme?: Scheme, options?: ParseOptions): UpcomingResult {
  const { ast } = parse(text, options);
  const items: UpcomingItem[] = [];

  // Find all items with due dates
  for (const item of ast.items) {
    if (item.metadata.due) {
      const upcomingItem: UpcomingItem = {
        id: item.id,
        due: item.metadata.due,
        priority: item.metadata.priority,
        path: buildPath(item.id, ast),
        item_span: item.item_span,
      };

      // Add priority label if scheme is provided
      if (scheme && item.metadata.priority !== null) {
        const label = scheme.priorities[String(item.metadata.priority)];
        if (label) {
          upcomingItem.priority_label = label;
        }
      }

      items.push(upcomingItem);
    }
  }

  // Map id -> sort-anchor (due_start if present, else due). Span items sort by
  // when they start mattering; point items sort by their date as before.
  const sortAnchor = new Map<string, string>();
  for (const item of ast.items) {
    if (item.metadata.due) {
      sortAnchor.set(item.id, item.metadata.due_start ?? item.metadata.due);
    }
  }

  // Sort by:
  // 1. due_start ?? due ascending (windows surface from when they start)
  // 2. Priority ascending (lower is higher priority, nulls last)
  // 3. Document order (by item_span start)
  items.sort((a, b) => {
    const dateCompare = sortAnchor.get(a.id)!.localeCompare(sortAnchor.get(b.id)!);
    if (dateCompare !== 0) return dateCompare;

    // Priority comparison (null treated as infinity)
    const aPri = a.priority ?? Infinity;
    const bPri = b.priority ?? Infinity;
    if (aPri !== bPri) return aPri - bPri;

    // Document order
    return a.item_span[0] - b.item_span[0];
  });

  return { items };
}

function parseMiscLocation(misc: string): { filename: string; heading: string } {
  const slashIndex = misc.indexOf('/');
  if (slashIndex === -1) {
    return { filename: misc, heading: 'Misc' };
  }
  return {
    filename: misc.substring(0, slashIndex),
    heading: misc.substring(slashIndex + 1),
  };
}

export function queryMisc(text: string, scheme?: Scheme): MiscResult {
  const { ast } = parse(text);
  const items: MiscItem[] = [];

  // Determine the heading name from scheme or use default
  const miscLocation = parseMiscLocation(scheme?.misc ?? 'todoosy.md/Misc');

  // Find the Misc section
  let miscSectionId: string | null = null;
  for (const item of ast.items) {
    if (item.type === 'heading' && item.title_text === miscLocation.heading && item.level === 1) {
      miscSectionId = item.id;
      break;
    }
  }

  if (miscSectionId === null) {
    return { items };
  }

  // Get all direct children of the Misc section
  const miscSection = ast.items.find(i => i.id === miscSectionId);
  if (!miscSection) {
    return { items };
  }

  const itemMap = new Map<string, ItemNode>();
  for (const item of ast.items) {
    itemMap.set(item.id, item);
  }

  for (const childId of miscSection.children) {
    const child = itemMap.get(childId);
    if (child) {
      items.push({
        id: child.id,
        title_text: child.title_text,
        item_span: child.item_span,
      });
    }
  }

  return { items };
}

export interface HashtagItem {
  id: string;
  title_text: string;
  path: string;
  item_span: [number, number];
  hashtags: string[];           // Direct hashtags
  effective_hashtags: string[]; // Including inherited
}

export interface HashtagResult {
  items: HashtagItem[];
}

export interface HashtagListResult {
  hashtags: string[];
}

export function queryByHashtag(text: string, hashtag: string): HashtagResult {
  const { ast } = parse(text);
  const items: HashtagItem[] = [];

  // Normalize hashtag (lowercase, remove # if present)
  const normalizedHashtag = hashtag.replace(/^#/, '').toLowerCase();

  // Find all items with the hashtag (using effective_hashtags for inheritance)
  for (const item of ast.items) {
    if (item.metadata.effective_hashtags.includes(normalizedHashtag)) {
      items.push({
        id: item.id,
        title_text: item.title_text,
        path: buildPath(item.id, ast),
        item_span: item.item_span,
        hashtags: item.metadata.hashtags,
        effective_hashtags: item.metadata.effective_hashtags,
      });
    }
  }

  // Sort by document order
  items.sort((a, b) => a.item_span[0] - b.item_span[0]);

  return { items };
}

export function listHashtags(text: string): HashtagListResult {
  const { ast } = parse(text);
  const hashtagSet = new Set<string>();

  // Collect all unique hashtags from all items
  for (const item of ast.items) {
    for (const tag of item.metadata.hashtags) {
      hashtagSet.add(tag);
    }
  }

  // Return sorted list
  return { hashtags: [...hashtagSet].sort() };
}
